### ISSUE TITLE
Question about big-O complexity

### ISSUE THREAD

**Issue by leijurv** [btrekkie/dynamic-connectivity#4]
I'm looking to rely on dynamic connectivity in a project I'm working on, which is a Minecraft mod to automatically construct schematics ingame. My mod already has a semi-working version of schematic construction but I am sprucing it up with a total rewrite. While the mod is popular (5 million downloads https://github.com/cabaletta/baritone/), a constant and chronic pain point for many years has been that the schematic builder currently makes poor decisions and doesn't plan ahead well enough. The basic problem is that while building a schematic, the surface that the player is walking around on is constantly changing (the player generally is walking around on top of what it's building, while it's building it). In order to efficiently answer the question "is it okay to place this block? or would that get me stuck and unable to walk onwards?", I see dynamic connectivity is the obvious choice. As blocks are placed and broken in the world, the edges representing where the player can walk are also created and destroyed. We should avoid placing a block that would split the walkable area into two disconnected components, and dynamic connectivity can efficiently answer the question of whether that would happen. (and another reason to use euler tour forests for this is that eventually I intend to add interesting augmentations (e.g. "do I still have remaining blocks that I intend to eventually place in this area, or is this area all done" could be an augmentation, then when areas are cut off from each other, it can realize "actually this is okay, there's no need to go to that side anymore (the component augmentation says none of the nodes have any more work to do) so it's okay to cut it off, just need to make sure I'm standing on the alive side when i do it"))

I've been looking into hopefully using your implementation of euler tour forests, as I gave up on my attempt about a year ago (https://github.com/cabaletta/baritone/commit/236d171d151d44fafae12426043f8667380f1a84) and lost interest. I got stuck on trying to get the big-O complexity down. Specifically, because I needed a binary tree whose split and concatenate were actually `O(log n)` and not `O((log n)^2)`. I was using a splay tree (https://github.com/cabaletta/baritone/commit/236d171d151d44fafae12426043f8667380f1a84#diff-849980cc61d114cb8bfda53f8ab63d972702b3a66a8f5fa28e9eb78cd765d02bR88-R184), because it's easy to implement while still having amortized approximate `O(log n)` split and concat, which is exactly what we want, but the worst case is `O(n)`. See https://en.wikipedia.org/wiki/Splay_tree In practice, I found that the tree structure could become highly unbalanced. I can't link to the test case because it's deep in an old repo that didn't quite make the cut, but it was basically the exact same as the basic performance tests I have of your dynamic connectivity: https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/test/java/baritone/builder/connectivity/ConnGraphTest.java#L44-L140 Essentially I just copy pasted the tests that my old code struggled with, and yours did great! In short, it creates a large grid, then cuts off the center. In my code, when run normally, it was quite slow, and I debugged it to realize that the average height of each forest was about 150 (splay trees don't guarantee height of the tree scales with `O(log n)`, that's only an approximation). I added some code to randomly splay each node in the tree, then run the operation, then the average height was decreased to 6, which made it actually have the correct big-O notation. In summary, the splay tree doesn't have strong enough time complexity guarantees in practice, and I suspect Minecraft's grid structure hits its worst-case scenario too often.

Since your code was promising there, I started working on on actual test of what I described, which is a "navigable surface" (meaning: the area where the Minecraft player can walk) that is updated in real time. https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/test/java/baritone/builder/NavigableSurfaceTest.java If you scroll through this you can see that it creates some interesting shapes! (the constraint is generally "place as many blocks as you can without ever cutting off your path back to the starting place", which results in cool outcomes that you can see in the tests (in my opinion the coolest one is https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/test/java/baritone/builder/NavigableSurfaceTest.java#L807-L830 ))

In terms of credit I have placed a comment crediting you and mentioning the license at the top of each file, there is a readme explaining where the implementation came from as well as the license (see https://github.com/cabaletta/baritone/tree/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie), and I used `git merge --allow-unrelated-histories` so that your commit history is not lost (so, for example, git blame still shows you as the author, for example see https://github.com/cabaletta/baritone/blame/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/connectivity/ConnGraph.java). I hope this is sufficient and makes it so it doesn't feel like I'm stepping on your toes!

Finally on to why I'm filing this issue.

I have been running your code through a profiler in my use case of Minecraft navigable surfaces and I fear I might be seeing an additional `O(log n)` factor. I hope I'm wrong! I'm seeing `removeEdge` generally take longer than `addEdge`, but that's fine since `addEdge` is actually `O(log n)` in practice, and the amortization is only to "pay for" future `removeEdge` calls. But, both are calling `RedBlackNode.split`, and `removeEdge` does it up to `O(log n)` times. And, my profiler says `split` is taking many times longer than `concatenate` (seven times longer, which is almost exactly the height of my tree (since it has about 100 elements in these tests generally)), which is what worried me. Reading through their implementations, I see that `concatenate` is not doing anything that worries me https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/red_black_node/RedBlackNode.java#L790-L885 but I think `split` might be. https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/red_black_node/RedBlackNode.java#L911-L1121 I think `split` is calling a `O(height)` function, `O(height)` times, when ideally the comment claims `split` should be `O(log n)` https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/red_black_node/RedBlackNode.java#L916 so it should not be permitted to do `O(height^2)` operations. Specifically, I think this while loop https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/red_black_node/RedBlackNode.java#L971 runs `O(log n)` times, while it calls https://github.com/cabaletta/baritone/blob/b2ffa7de1bcfa64f513c0b4b48f96e61e25605d0/src/main/java/baritone/builder/utils/com/github/btrekkie/red_black_node/RedBlackNode.java#L1013 which I fear might loop on a node's parent which could be up to `O(height)` loops? I can't quite tell what's going on - on the one hand I see that it sometimes breaks and also it only loops if the node is red, which makes me think it would only go `O(1)` iterations, but on the other hand I see that node is reassigned to be either its parent or its grandparent in some scenarios? **All in all, I can't quite clearly tell whether `fixInsertionWithoutGettingRoot` is `O(1)` or `O(height)`, this is the main point of the issue that I hope you can answer?** I fear `O(height)` because of how `insert` calls it, as I understand binary search trees it's the case that whenever you want to insert or delete a node, assuming you know where in the tree that operation will take place, the operation will take `O(log n)` amortized at minimum? And in your code it appears that that fixup is indeed `fixInsertionWithoutGettingRoot` so I fear it may be an `O(log n)` function. :(

<img width="770" alt="Screen Shot 2023-03-31 at 10 57 04 PM" src="https://user-images.githubusercontent.com/3837873/229140466-a5e306de-d2c7-40c1-86d4-44febe8fc2f3.png">

**tldr:** I worry RedBlackNode.split might be `O((log n)^2)` when it should be `O(log n)`, and I worry this would make ConnGraph.removeEdge be `O((log n)^3)` when it should be `O((log n)^2)`

Getting a binary search tree with actual reliable `O(log n)` split and concatenate as well as additional attachments (augmentations that are rolled up the tree) is actually why I gave up last year, it was just too intricate. Your implementation looks great and it's well commented and everything! Back then I was looking at https://stackoverflow.com/questions/3176863/concatenating-red-black-trees which seemed to indicate to me that some kind of modification was needed to get both to `O(log n)`? Also I was considering AVL trees which might be easier, for example OCaml uses them https://cs.stackexchange.com/questions/43214/split-in-avl-tree-with-complexity-o-log-n https://github.com/ocaml/ocaml/blob/b94c4b840fbf7af72157188e73cdb1a0a7fb8bb6/stdlib/set.ml#L201 If red black trees are too complicated in practice or too slow of a constant term or big-O complexity in practice I think I will attempt to find or implement myself perhaps an AVL tree to get the desired time bounds.

All that being said, I hope I'm wrong and there is no time complexity issue, or if there is one, it's easily fixable. 

Thanks for reading! Please let me know what you think.

**Comment by btrekkie** [comment]
I'm glad that you're finding this library useful, and even digging into the internals. I've enjoyed watching the changes you're making in your fork, and the mod you're working on sounds neat. Those are some interesting shapes that emerged out of your tests. (Incidentally, I also programmed something Minecraft-related, although it's not nearly as popular: https://www.paperpickaxe.com/ .)

It's surprisingly hard to find resources on how to split a red-black tree in O(log N) time. I can't find the original diagram I used to implement `split`, but thankfully, I can rely on the comment I wrote describing the implementation ( https://github.com/btrekkie/RedBlackNode/blob/1c8ccd4646e483c2e48ecd56a4bd91cf6c3b3695/src/main/java/com/github/btrekkie/red_black_node/RedBlackNode.java#L889-L906 ). (Also, I remember that the diagram didn't prove that it takes O(log N) time. When I implemented `split`, I was also worried about the possibility of an O(log^2 N) running time and had to convince myself that it was in fact O(log N). This issue gives me the opportunity to write out a formal proof of this claim and make me feel more confident about `RedBlackNode`.)

As you suggested, `split` makes O(log N) calls to `fixInsertionWithoutGettingRoot`, and `fixInsertion...` takes Θ(log N) time in the worst case. However, in this particular usage, the calls to `fixInsertion...` should only take O(log N) time in total. We can show that the fixups in the post-split tree take O(log N) time using something akin to the potential method of amortized analysis, where the potential function P is equal to the number of red nodes that are right children of nodes in the path from `last` to the root.

Let's try and show that the calls to `fixInsertion...` for the post-split tree take O(log N) time total in a given call to `split`. (By symmetry, this implies that they take O(log N) time for the pre-split tree.) Initially, the potential function P is 0, since there is no post-split tree. Every time P increases by 1, we'll charge the current operation for work that will (possibly) be performed by some future operation that decreases P by 1.

There are four types of operations that affect the structure of the post-split tree, the colors of its nodes, or the value of `last`, and thus four types of operations that could possibly affect the value of P:

* Initializing the post-split tree.
* Moving `last` to its left child.
* Adding a pivot node.
* Calling `fixInsertion...`.

Let's compute the total cost of these four types of operations. Initializing the post-split tree (i.e. changing `last` from null to non-null) changes P from 0 to at most 1. (Since `last` is the root at that point, there is only one node in the path from `last` to the root, meaning there is at most one red right child.) So the cost of initializing the post-split tree (which occurs at most once) is O(1), plus up to 1 for the increase in P.

Moving `last` to its left child increases P by at most 1, since it adds exactly one node to the collection of nodes that are right children of the `last`-to-root path. So the cost of each move is O(1), plus up to 1 for the increase in P. There are O(log N) such moves, because the initial height of `last` is O(log N), so the total cost is O(log N) * O(1) = O(log N).

Adding a pivot node (sticking `lastPivot` in the left spine and giving it two children) adds exactly two nodes to the collection of nodes in the `last`-to-root path: the pivot node itself and its left child. (It also removes one node from this collection: its right child.) So it increases P by at most 2 (actually, at most 1, since we know that one of the right children is black), and the cost is at most O(1) + 2. We add pivot nodes O(log N) times (up to once for each node in the path from the root of the original tree to the split node). So the total cost for adding pivot nodes is O(log N) * O(1) = O(log N).

For insertion fixup, let's work off of the description depicted in https://en.wikipedia.org/wiki/Red%E2%80%93black_tree#Insertion , since that has some nice diagrams. (Just recognize that in `RedBlackNode`, we also need to color the root black in case I3, because unlike in the Wikipedia article, `RedBlackNode` requires the root to be black.) Note that the only problematic case is I2, because that's the only case where we can't finish in a constant amount of time; we need to progress up higher in the tree.

Let's look at a call to `fixInsertion...`. The cost for cases I1, I3, and I4 is O(1), and none of them affects P. It is impossible to reach case I5, because the current node is always on the left spine of the tree. Case I6 increases P by 1, since the grandparent becomes a new red right child of a node in the `last`-to-root path. So the cost of case I6 is O(1) + 1. Case I2 decreases P by 1, since it changes the uncle from red to black. So the cost for case I2 is 1 - 1 = 0. The operation is free, meaning it is charged to some previous operation (either initializing the post-split tree, moving `last` to its left child, adding a pivot node, or reaching case I6 during a previous call to `fixInsertion...`).

The total cost of the call to `fixInsertion...` is the total cost for all 6 cases. Case I2 is reached O(log N) times, and the other cases are reached at most once, so the total cost is O(1) + O(log N) * 0 + O(1) + O(1) + 0 + O(1) = O(1). There are at most O(log N) calls to `fixInsertion...` (at most one for each node in the path from the root of the original tree to the split node). Thus, the total running time for all four types of operations is O(1) + O(log N) + O(log N) + O(log N) * O(1) = O(log N).

In fact, I think we can sort of empirically verify the claim that the calls to `fixInsertion...` take O(log N) time in total. I believe that the calls to `fixInsertion...` will visit each individual node at most a constant number of times: 2 or possibly 3. (This doesn't follow from the above proof, but from a different line of reasoning. I haven't worked this up into a formal proof, but I have this idea that each time we reach case I2 or at least reach it multiple times in a row, it destroys our future ability to reach case I2 again at the same part of the left spline.)

You can empirically check this in your test cases by modifying `RedBlackNode`. The idea would be to change (or rather override) `fixInsertion...` to take as a parameter a map from each node to the number of times we have visited the node. At the beginning of the while loop, it would increment the visit count for "node". Then at the end of the `split` method, we could check the total visit counts for each node during the course of the call to `split`. If I'm correct, the visit count for a node should never exceed 3 (or perhaps even 2). So if you buy that `fixInsertion...` visits each node a constant number of times throughout the course of a call to `split`, and if you buy that each iteration of the while loop in `fixInsertion...` takes a constant amount of time, and if you buy that the calls to `fixInsertion...` visit O(log N) different nodes in all (which is actually not so obvious, oh well), then you should buy that the calls to `fixInsertion...` take O(log N) time in total.

We also need to be careful in the rest of the implementation of `split` to avoid O(log^2 N) running time. For example, we use `first` and `last` variables to avoid repeatedly walking down from the roots of the trees, and we're careful not to repeatedly walk up to the root of the tree when calling `augment()`. But I think we're good here, and everything else takes O(log N) time as well.

Please let me know if you find an error in my reasoning or a bug in my implementation.

Regarding the Stack Overflow page you linked, I haven't dug into it too deeply, but here's what I noticed from a quick glance. The accepted answer https://stackoverflow.com/a/3224033/10935386 refers to a "red-black tree with fingers" data structure, which takes amortized O(log(min(p, q))) time to split into trees of size p and q. So I think the article was suggesting that red-black trees need to be modified in some special way to achieve O(log(min(p, q))) split time, but it was not necessarily suggesting that they need to be modified to achieve O(log N) split time.

If you do end up trying AVL trees, I would be curious to know how they compare to red-black trees in terms of performance. I'm surprised that splay trees weren't performant. According to https://en.wikipedia.org/wiki/Splay_tree , it's conjectured that the amortized performance of splay trees is within a constant factor of optimal, regardless of the access pattern. I wonder if maybe there was a bug in your implementation. I know that in the past when I implemented a splay tree, I struggled with making the appropriate splays, and unlike in red-black trees, nothing would obviously break when I messed it up. And working with uncommon operations that are hard to find online explanations for, like split and concatenate, could definitely make things more challenging. In any event, it shouldn't be **too** hard to change `ConnGraph` to use a different kind of binary search tree, provided you try to mimic the interface of `RedBlackNode` (or even create a shared superclass for `RedBlackNode` and whatever other kind of tree node).

I'll be curious to know what you discover about the real-world performance of `ConnGraph` and `RedBlackNode`. And maybe once it's integrated into your mod, you could send me some realistic sample sequences of operations that I could use to optimize `ConnGraph`. Depending on my time and motivation, I could finally try implementing some of the ideas I listed in https://github.com/btrekkie/dynamic-connectivity/blob/master/optimization_ideas.txt .

**Comment by btrekkie** [comment]
I noticed something strange in the profiler screenshot you posted. It seems to suggest that `split` called `fixInsertion...`, which then directly called `augment()`. But this should be impossible, because the `augment` parameter that `split` passes to `fixInsertion...` is false. `fixInsertion...` should only call `augment()` indirectly, through a call to `rotateLeft()` or `rotateRight()`. I wonder, is this a result of changes you made to `RedBlackNode`, a quirk or error in the profiler, or a bug in `fixInsertion...`, or am I misreading the profiler results?

**Comment by leijurv** [comment]
> I'm glad that you're finding this library useful, and even digging into the internals. I've enjoyed watching the changes you're making in your fork, and the mod you're working on sounds neat. Those are some interesting shapes that emerged out of your tests. 

Yay! Paper Pickaxe looks super cool too!

> It's surprisingly hard to find resources on how to split a red-black tree in O(log N) time.

That was my experience exactly.

With respect to your explanation, thank you so much, that's very comprehensive and thorough! I believe you :)

> If you do end up trying AVL trees, I would be curious to know how they compare to red-black trees in terms of performance.

Will do, but switching to AVL is probably a low priority :)

> According to https://en.wikipedia.org/wiki/Splay_tree , it's conjectured that the amortized performance of splay trees is within a constant factor of optimal, regardless of the access pattern. I wonder if maybe there was a bug in your implementation.

Possibly, although it's very simple https://github.com/cabaletta/baritone/blob/af320bd0527398d2fe7500152f13d134a81667e4/src/main/java/baritone/builder/EulerTourForest.java#L121-L143

> And working with uncommon operations that are hard to find online explanations for, like split and concatenate, could definitely make things more challenging.

Split and concat are actually pretty simple in splay trees https://github.com/cabaletta/baritone/blob/af320bd0527398d2fe7500152f13d134a81667e4/src/main/java/baritone/builder/EulerTourForest.java#L145-L187 you just splay a specific node, then connecting up the roots becomes easy. For example to concat you can splay the highest node of the left tree, then you set it's right child to the right tree.

> In any event, it shouldn't be too hard to change ConnGraph to use a different kind of binary search tree, provided you try to mimic the interface of RedBlackNode (or even create a shared superclass for RedBlackNode and whatever other kind of tree node).

Might do this just for fun! Perhaps the simplicity of the splay tree could help in practice on real schematics, even though I found it to be net negative on the grid test case.

> I'll be curious to know what you discover about the real-world performance of ConnGraph and RedBlackNode. And maybe once it's integrated into your mod, you could send me some realistic sample sequences of operations that I could use to optimize ConnGraph. Depending on my time and motivation, I could finally try implementing some of the ideas I listed in https://github.com/btrekkie/dynamic-connectivity/blob/master/optimization_ideas.txt .

I'll keep you posted for sure! I was actually imagining this already - if I have a large schematic for a real world use case, I could somehow export the series of operations, which could be used for optimization for sure!


I've written down how the new builder will work https://github.com/cabaletta/baritone/blob/builder-2/src/main/java/baritone/builder/README.md#how-will-it-work I suspect the devil will be in the details, specifically, in nailing down the exact conditions for if/when it's allowed to split the navigable surface into two. But in any event, dynamic connectivity is going to be the core of the planner :)

**Comment by btrekkie** [comment]
Okay, I see at least two performance bugs in the splay tree-based implementation. The key thing is that whenever we "access" a node (for some definition of access), we need to splay it. The idea is that if there's a node that's overly deep, it might take a long time to access it, but at least when we do so, that portion of the tree will become more balanced for future operations.

In the call to `connected`, both of the nodes need to be splayed, e.g.:

```java
public boolean connected(int vertA, int vertB) {
    BSTNode nodeA = loopbacks[vertA];
    BSTNode nodeB = loopbacks[vertB];
    nodeA.splay();
    boolean connected = nodeB.walkAncestor() == nodeA;
    nodeB.splay();
    return connected;
}
```

At least, I **think** that's what's needed. I only learned splay trees in the context of accessing nodes by going down from the root, as opposed to jumping directly to a node and then going up to the root.

Also, I realize this code is just for debugging, but https://github.com/cabaletta/baritone/blob/af320bd0527398d2fe7500152f13d134a81667e4/src/main/java/baritone/builder/EulerTourForest.java#L75-L80 is problematic in that we're accessing the min and loopback nodes without splaying them (I think).

It looks like you were close though!

That write-up is interesting and looks well thought out. Initially I was wondering if there was some special algorithm that could solve the planning problem optimally, rather than a more greedy/ad-hoc type approach, but now I'm thinking that probably every formulation of the problem is NP-hard and so not really feasible. (For example, placing N blocks in specific locations in a plane with a minimum amount of walking is similar to the traveling salesman problem.) Anyway, I'm sure you've thought about this problem much more deeply than I have.

**Comment by leijurv** [comment]
I'm sure that would help in practice, but, in the use case that I was benching, all I was doing was creating a grid. Thank you for reminding me that connected should splay too, but I didn't get to the point where it was being called (except for the trivial calls you can see in that file).

Since splay trees are so simple I will keep them on the table and once I have a realistic access pattern I'll check again. It's easy for me to believe that in practice the accesses from the planner checking for connectedness could keep the tree roughly balanced.

Edit: I just recreated a similar enough test in the repo: https://github.com/cabaletta/baritone/commit/c9bfcb417e7647bc91b2fb7f46945ef09ff1b4ab and yes it does appear that splaying upon each connected check is pretty good - notably slower than running connection checks on an already balanced tree, but also notably faster than running all the connection checks on an unbalanced tree.


Yeah I came to the conclusion that any kind of search over the space of block placements is impractical. Even simple builds have hundreds of blocks, and it's 2^blocks possible states. So any kind of backtracking search is a no-go, therefore, heuristics / greedy approach.
